Automattic/jetpack
# AI Featured Image: experiment with Stable Diffusion generation

This is a proof-of-concept for using Stable Diffusion on the AI Featured Image Generator instead of Dall-e 3.

## Proposed changes:

* Create a feature flag and a filter to control the visibility of the proof-of-concept
* Create a function to call the Jetpack AI completion endpoint on a sync, non-streamed way
* Support image generation using Stable Diffusion on the `useImageGenerator` hook
   * Use GPT to pre-process the post content and the user prompt to generate a valid Stable Diffusion prompt
   * use the GPT output as input for the Stable Diffusion generation
   * keep the return value in the same format as the Dall-e one, with Base64 image data and the revised (GPT-generated) prompt on the return value
* Support different costs between the default, dall-e powered generation, and the experimental, stable diffusion powered generation (using cost = 1)

Known problems:

- there is a flick in the modal height when changing between a generated image and the loading state; this happens because Stable Diffusion images follows the 4:3 ratio but the placeholder of the loading state has the 1.75:1 ratio, used on the Dall-e images; since we plan to revamp the empty/loading state in the near future, this seems like a good thing to ignore here
- users may be charged one request when the image generation fails; since we use GPT (that costs 1 request) and then Stable Diffusion (that costs 0), if the Stable Diffusion call fails, the user will still be charged for the GPT request even without an image being generated; we can fix it by changing the way we charge the user (charge 0 for the GPT call and then 1 for the Stable Diffusion call), but I left it out since it's a backend change, out of the scope of a PoC

### Other information:


## Jetpack product discussion



## Does this pull request change what data or activity we track or use?



## Testing instructions:






**Preparing the testing site:**

* Run this branch on your testing site (a fresh JN site, your local site, or a sandboxed Simple site)
* Make sure Jetpack is connected, if you are testing on a Jetpack site
* Switch the block variation to `beta`:
   * On JN sites, go to `Settings > Jetpack Constants > Type of blocks to load in the editor > Beta`
   * On your local, development site, add `define( 'JETPACK_BLOCKS_VARIATION', 'beta' );` on the `wp-config.php` file
   * On a Simple site, use the site while proxied
* Enable the `jetpack_experimental_image_generation_enabled` filter using `add_filter( 'jetpack_experimental_image_generation_enabled', '__return_true' );`
   * On JN sites, use the Code Snippets plugin to add the code as a new snippet
   * On your local, development site, add the snippet to the bottom of the `wp-config.php` file
   * On a Simple site, add the snippet to the `0-sandbox.php` file

**Testing:**

* Open the block editor and add some content to it (tip: use the AI Assistant to generate a few paragraphs about some topic)
* Look for the AI Featured Image tool (can be the sidebar button, or the media source dropdown menu on the featured image section) and open it
* Confirm the tool will show a new image after some seconds
* Confirm you can type a custom prompt and steer the image generation
* Confirm the images are good/usable and related to the content or to the prompt you type
* Confirm each image is costing 1 request by watching the request balance
* Confirm you can set the image as featured and it will show up on the post
* Confirm all the images you generate are automatically saved to the site's media library
